decorators_library/decorators.py

```diff
@@ -1 +1,132 @@
+            raise TimeoutError('Function call timed out')
+        return return_val
+
+def timeout(max_time):
+    def decorator(func):
+        return Timeout(func, max_time)
+    return decorator
+    
+def debug(logger=None):
+    def other(func):
+        def wrapper(*args, **kwargs):
+            return_val = func(*args, **kwargs)
+            if logger is not None:
+                log = logger
+            else:  
+                log = getLogger('tests.test_decorators')
+                
+            log.debug('Executing "{}" with params: {}, {{}}'.format(func.__name__, tuple(args)))
+            log.debug('Finished "{}" execution with result: {}'.format(func.__name__, return_val))
+                
+            return return_val
+        return wrapper
+    return other
+  
+class count_calls(object):
+    
+    func_calls = {}
+    
+    def __init__(self,func):
+        self.func = func
+        self.func_name = func.__name__
+        self.func_calls.setdefault(self.func_name, 0)
+        
+
+    def counter(self):
+        return self.func_calls[self.func_name]
+        
+    @classmethod
+    def counters(self):
+        return self.func_calls
+        
+    @classmethod
+    def reset_counters(self):
+        self.func_calls = {}
+    
+    def __call__(self, *args, **kwargs):
+        self.func_calls[self.func_name] += 1
+        
+        return self.func(*args, **kwargs)
+    
+class memoized:
+    def __init__(self, func):
+        self.func = func
+        self.cache = {}
+        
+    def __call__(self, *args, **kwargs):
+        return_val = self.func(*args, **kwargs)
+        call_dict = {tuple(args) : return_val}
+        self.cache.update(call_dict)
+        return return_val
+        
+    
+def assert_type(*args, **kwargs):
+    """Decorator that does type checking for each parameter."""
+    type_list = list(args)
+    type_dict = kwargs
+    
+    def decorator(func):
+        def wrapper(*args, **kwargs):
+            
+            if len(type_list) + len(type_dict) != len(args) + len(kwargs):
+                expected = len(inspect.getargspec(func).args)
+                received = len(args) + len(kwargs)
+                raise ValueError("Wrong number arguments. Expected: {}; Received: {}".format(expected, received))
+                
+            type_args = [type(val) for val in args]
+            type_kwargs = [kwargs[key] for key in type_dict] 
+            
+            if type_args != type_list or not all([type_dict[key] == kwargs[key] for key in type_dict]):
+                expected = tuple(type_list + [type_dict[key] for key in type_dict])
+                received = tuple(type_args + type_kwargs)
+                error_msg = "Wrong type. Expected: {}; Received: {}".format(get_type(expected), get_type(received))
+                
+                raise TypeError(error_msg) 
+                
+            return func(*args, **kwargs)
+        return wrapper
+    return decorator
+
```

Comment:
smart idea for a decorator! seems like it would be really helpful in some of our projects haha